<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Agglomerate data using taxonomic information — agglomerate-methods • mia</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Agglomerate data using taxonomic information — agglomerate-methods"><meta property="og:description" content="agglomerateByRank can be used to sum up data based on the association
to certain taxonomic ranks given as rowData. Only available
taxonomyRanks can be used."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/microbiome/mia/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Agglomerate data using taxonomic information</h1>
    <small class="dont-index">Source: <a href="https://github.com/microbiome/mia/blob/HEAD/R/agglomerate.R" class="external-link"><code>R/agglomerate.R</code></a></small>
    <div class="hidden name"><code>agglomerate-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>agglomerateByRank</code> can be used to sum up data based on the association
to certain taxonomic ranks given as <code>rowData</code>. Only available
<code><a href="taxonomy-methods.html">taxonomyRanks</a></code> can be used.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for SummarizedExperiment</span></span>
<span><span class="fu">agglomerateByRank</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  rank <span class="op">=</span> <span class="fu"><a href="taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  onRankOnly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  na.rm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  empty.fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="st">""</span>, <span class="st">" "</span>, <span class="st">"\t"</span>, <span class="st">"-"</span>, <span class="st">"_"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SingleCellExperiment</span></span>
<span><span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, altexp <span class="op">=</span> <span class="cn">NULL</span>, strip_altexp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for TreeSummarizedExperiment</span></span>
<span><span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, agglomerateTree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code>
object</p></dd>


<dt>rank</dt>
<dd><p>a single character defining a taxonomic rank. Must be a value of
<code><a href="taxonomy-methods.html">taxonomyRanks()</a></code> function.</p></dd>


<dt>onRankOnly</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: Should information only from
the specified rank be used or from ranks equal and above? See details.
(default: <code>onRankOnly = FALSE</code>)</p></dd>


<dt>na.rm</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: Should taxa with an empty rank be
removed? Use it with caution, since empty entries on the selected rank
will be dropped. This setting can be tweaked by defining
<code>empty.fields</code> to your needs. (default: <code>na.rm = TRUE</code>)</p></dd>


<dt>empty.fields</dt>
<dd><p>a <code>character</code> value defining, which values should be
regarded as empty. (Default: <code>c(NA, "", " ", "\t")</code>). They will be
removed if <code>na.rm = TRUE</code> before agglomeration.</p></dd>


<dt>...</dt>
<dd><p>arguments passed to <code>agglomerateByRank</code> function for
<code>SummarizedExperiment</code> objects,
<code><a href="merge-methods.html">mergeRows</a></code> and
<code><a href="https://rdrr.io/pkg/scuttle/man/sumCountsAcrossFeatures.html" class="external-link">sumCountsAcrossFeatures</a></code>.</p><ul><li><p><code>remove_empty_ranks</code>A single boolean value for selecting
whether to remove those columns of rowData that include only NAs after
agglomeration. (By default: <code>remove_empty_ranks = FALSE</code>)</p></li>
<li><p><code>make_unique</code>A single boolean value for selecting
whether to make rownames unique. (By default: <code>make_unique = TRUE</code>)</p></li>
</ul></dd>


<dt>altexp</dt>
<dd><p>String or integer scalar specifying an alternative experiment
containing the input data.</p></dd>


<dt>strip_altexp</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: Should alternative
experiments be removed prior to agglomeration? This prevents to many
nested alternative experiments by default (default:
<code>strip_altexp = TRUE</code>)</p></dd>


<dt>agglomerateTree</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>: should
<code>rowTree()</code> also be agglomerated? (Default:
<code>agglomerateTree = FALSE</code>)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A taxonomically-agglomerated, optionally-pruned object of the same
class as <code>x</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Based on the available taxonomic data and its structure setting
<code>onRankOnly = TRUE</code> has certain implications on the interpretability of
your results. If no loops exist (loops meaning two higher ranks containing
the same lower rank), the results should be comparable. you can check for
loops using <code><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/detectLoop.html" class="external-link">detectLoop</a></code>.</p>
<p>Agglomeration sum up values of assays at specified taxonomic level. Certain assays,
e.g. those that include binary or negative values, can lead to meaningless values,
when values are summed. In those cases, consider doing agglomeration first and then
transformation.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="merge-methods.html">mergeRows</a></code>,
<code><a href="https://rdrr.io/pkg/scuttle/man/sumCountsAcrossFeatures.html" class="external-link">sumCountsAcrossFeatures</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># print the available taxonomic ranks</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</span>
<span class="r-in"><span><span class="fu"><a href="taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># agglomerate at the Family taxonomic rank</span></span></span>
<span class="r-in"><span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">GlobalPatterns</span>, rank<span class="op">=</span><span class="st">"Family"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## How many taxa before/after agglomeration?</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 19216</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 603</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># with agglomeration of the tree</span></span></span>
<span class="r-in"><span><span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">GlobalPatterns</span>, rank<span class="op">=</span><span class="st">"Family"</span>,</span></span>
<span class="r-in"><span>                       agglomerateTree <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The root is added with label 'ALL'</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span> <span class="co"># same number of rows, but</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 603</span>
<span class="r-in"><span><span class="fu">rowTree</span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="co"># ... different</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylogenetic tree with 19216 tips and 19215 internal nodes.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tip labels:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   549322, 522457, 951, 244423, 586076, 246140, ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Node labels:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Rooted; includes branch lengths.</span>
<span class="r-in"><span><span class="fu">rowTree</span><span class="op">(</span><span class="va">x2</span><span class="op">)</span> <span class="co"># ... tree</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylogenetic tree with 496 tips and 260 internal nodes.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tip labels:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Family:Cenarchaeaceae, Family:SAGMA-X, Family:Nitrososphaeraceae, Family:Sulfolobaceae, Family:Halobacteriaceae, Family:MSBL1, ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Node labels:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Rooted; includes branch lengths.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co"># If assay contains binary or negative values, summing might lead to meaningless</span></span></span>
<span class="r-in"><span> <span class="co"># values, and you will get a warning. In these cases, you might want to do </span></span></span>
<span class="r-in"><span> <span class="co"># agglomeration again at chosen taxonomic level.</span></span></span>
<span class="r-in"><span> <span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="transformCounts.html">transformCounts</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, method <span class="op">=</span> <span class="st">"pa"</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>'pa' includes binary values.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Agglomeration of it might lead to meaningless values.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Check the assay, and consider doing transformation again manually with agglomerated data.</span>
<span class="r-in"><span> <span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="transformCounts.html">transformCounts</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"pa"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># removing empty labels by setting na.rm = TRUE</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span><span class="op">$</span><span class="va">Family</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 5603</span>
<span class="r-in"><span><span class="va">x3</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">GlobalPatterns</span>, rank<span class="op">=</span><span class="st">"Family"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x3</span><span class="op">)</span> <span class="co"># different from x2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 341</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Because all the rownames are from the same rank, rownames do not include </span></span></span>
<span class="r-in"><span><span class="co"># prefixes, in this case "Family:". </span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x3</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Sulfolobaceae"  "SAGMA-X"        "Cenarchaeaceae"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To add them, use getTaxonomyLabels function.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">x3</span>, with_rank <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x3</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Family:Sulfolobaceae"  "Family:SAGMA-X"        "Family:Cenarchaeaceae"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># use 'remove_empty_ranks' to remove columns that include only NAs</span></span></span>
<span class="r-in"><span><span class="va">x4</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">GlobalPatterns</span>, rank<span class="op">=</span><span class="st">"Phylum"</span>, remove_empty_ranks <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">x4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 2 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           Kingdom         Phylum</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       &lt;character&gt;    &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:Crenarchaeota      Archaea  Crenarchaeota</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:Euryarchaeota      Archaea  Euryarchaeota</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:Actinobacteria    Bacteria Actinobacteria</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:Spirochaetes      Bacteria   Spirochaetes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:MVP-15            Bacteria         MVP-15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylum:Proteobacteria    Bacteria Proteobacteria</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If assay contains NAs, you might want to consider replacing them since summing-up</span></span></span>
<span class="r-in"><span><span class="co"># NAs lead to NA</span></span></span>
<span class="r-in"><span><span class="va">x5</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></span>
<span class="r-in"><span><span class="co"># Replace first value with NA</span></span></span>
<span class="r-in"><span><span class="fu">assay</span><span class="op">(</span><span class="va">x5</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span></span>
<span class="r-in"><span><span class="va">x6</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">x5</span>, <span class="st">"Kingdom"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu">assay</span><span class="op">(</span><span class="va">x6</span><span class="op">)</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             CL3     CC1    SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea      NA    1248  28811      33      57      42     112     140     303</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria 862815 1134209 668698 1543418 2076419  718901  433782  186157 2000099</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          M11Tong LMEpi24M SLEpi20M  AQC1cm  AQC4cm  AQC7cm    NP2     NP3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea       30      131      145    4459   24692   28051   1826   43197</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria  100157  2117461  1217167 1163289 2332489 1671242 521808 1435768</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              NP5 TRRsed1 TRRsed2 TRRsed3   TS28    TS29   Even1  Even2   Even3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea    33996     843    8418   14250   1598    1690     150     23      91</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria 1618758   57845  484708  265454 935868 1209381 1215987 971050 1078150</span>
<span class="r-in"><span><span class="co"># Replace NAs with 0. It is justified when we are summing-up counts</span></span></span>
<span class="r-in"><span><span class="fu">assay</span><span class="op">(</span><span class="va">x5</span><span class="op">)</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">x5</span><span class="op">)</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></span>
<span class="r-in"><span><span class="va">x6</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">x5</span>, <span class="st">"Kingdom"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu">assay</span><span class="op">(</span><span class="va">x6</span><span class="op">)</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             CL3     CC1    SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea    1262    1248  28811      33      57      42     112     140     303</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria 862815 1134209 668698 1543418 2076419  718901  433782  186157 2000099</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          M11Tong LMEpi24M SLEpi20M  AQC1cm  AQC4cm  AQC7cm    NP2     NP3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea       30      131      145    4459   24692   28051   1826   43197</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria  100157  2117461  1217167 1163289 2332489 1671242 521808 1435768</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              NP5 TRRsed1 TRRsed2 TRRsed3   TS28    TS29   Even1  Even2   Even3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Archaea    33996     843    8418   14250   1598    1690     150     23      91</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteria 1618758   57845  484708  265454 935868 1209381 1215987 971050 1078150</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Look at enterotype dataset...</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## print the available taxonomic ranks. Shows only 1 rank available</span></span></span>
<span class="r-in"><span><span class="co">## not useful for agglomerateByRank</span></span></span>
<span class="r-in"><span><span class="fu"><a href="taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Genus"</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Sudarshan A. Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

