<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate weighted or unweighted (Fast) UniFrac distance — calculateUniFrac • mia</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculate weighted or unweighted (Fast) UniFrac distance — calculateUniFrac" />
<meta property="og:description" content="This function calculates the (Fast) UniFrac distance for all sample-pairs
in a TreeSummarizedExperiment
object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/microbiome/mia/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate weighted or unweighted (Fast) UniFrac distance</h1>
    <small class="dont-index">Source: <a href='https://github.com/microbiome/mia/blob/master/R/calculateUniFrac.R'><code>R/calculateUniFrac.R</code></a></small>
    <div class="hidden name"><code>calculateUniFrac.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates the (Fast) UniFrac distance for all sample-pairs
in a <code><a href='https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-class.html'>TreeSummarizedExperiment</a></code>
object.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for ANY,phylo</span>
<span class='fu'>calculateUniFrac</span><span class='op'>(</span>
  <span class='va'>x</span>,
  <span class='va'>tree</span>,
  weighted <span class='op'>=</span> <span class='cn'>FALSE</span>,
  normalized <span class='op'>=</span> <span class='cn'>TRUE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>SerialParam</span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='co'># S4 method for TreeSummarizedExperiment,missing</span>
<span class='fu'>calculateUniFrac</span><span class='op'>(</span><span class='va'>x</span>, exprs_values <span class='op'>=</span> <span class='st'>"counts"</span>, transposed <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>runUniFrac</span><span class='op'>(</span>
  <span class='va'>x</span>,
  <span class='va'>tree</span>,
  weighted <span class='op'>=</span> <span class='cn'>FALSE</span>,
  normalized <span class='op'>=</span> <span class='cn'>TRUE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>SerialParam</span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a numeric matrix or a
<code><a href='https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-class.html'>TreeSummarizedExperiment</a></code>
object containing a tree.</p>
<p>Please  note that <code>runUniFrac</code> expects a matrix with samples per row
and not per column. This is implemented to be compatible with other
distance calculations such as <code><a href='https://rdrr.io/r/stats/dist.html'>dist</a></code> as much as
possible.</p></td>
    </tr>
    <tr>
      <th>tree</th>
      <td><p>if <code>x</code> is a matrix, a
<code><a href='https://rdrr.io/pkg/TreeSummarizedExperiment/man/phylo.html'>phylo</a></code> object matching the
matrix. This means that the phylo object and the columns should relate
to the same type of features (aka. microorganisms).</p></td>
    </tr>
    <tr>
      <th>weighted</th>
      <td><p><code>TRUE</code> or <code>FALSE</code>: Should use weighted-UniFrac
calculation? Weighted-UniFrac takes into account the relative abundance of
species/taxa shared between samples, whereas unweighted-UniFrac only
considers presence/absence. Default is <code>FALSE</code>, meaning the
unweighted-UniFrac distance is calculated for all pairs of samples.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p><code>TRUE</code> or <code>FALSE</code>: Should the output be
normalized such that values range from 0 to 1 independent of branch length
values? Default is <code>TRUE</code>. Note that (unweighted) <code>UniFrac</code> is
always normalized by total branch-length, and so this value is ignored when
<code>weighted == FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>A
<code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html'>BiocParallelParam</a></code>
object specifying whether the UniFrac calculation should be parallelized.</p></td>
    </tr>
    <tr>
      <th>exprs_values</th>
      <td><p>a single <code>character</code> value for specifying which
assay to use for calculation.</p></td>
    </tr>
    <tr>
      <th>transposed</th>
      <td><p>Logical scalar, is x transposed with cells in rows?</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>optional arguments not used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a sample-by-sample distance matrix, suitable for NMDS, etc.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Please note that if <code>calculateUniFrac</code> is used as a <code>FUN</code> for
<code>runMDS</code>, the argument <code>ntop</code> has to be set to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(x)</a></code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><a href='http://bmf.colorado.edu/unifrac/'>http://bmf.colorado.edu/unifrac/</a></p>
<p>The main implementation (Fast UniFrac) is adapted from the algorithm's
description in:</p>
<p>Hamady, Lozupone, and Knight,
``<a href='http://www.nature.com/ismej/journal/v4/n1/full/ismej200997a.html'>Fast
UniFrac:</a> facilitating high-throughput phylogenetic analyses of microbial
communities including analysis of pyrosequencing and PhyloChip data.'' The
ISME Journal (2010) 4, 17--27.</p>
<p>See also additional descriptions of UniFrac in the following articles:</p>
<p>Lozupone, Hamady and Knight, ``UniFrac - An Online Tool for Comparing
Microbial Community Diversity in a Phylogenetic Context.'', BMC
Bioinformatics 2006, 7:371</p>
<p>Lozupone, Hamady, Kelley and Knight, ``Quantitative and qualitative (beta)
diversity measures lead to different insights into factors that structure
microbial communities.'' Appl Environ Microbiol. 2007</p>
<p>Lozupone C, Knight R. ``UniFrac: a new phylogenetic method for comparing
microbial communities.'' Appl Environ Microbiol. 2005 71 (12):8228-35.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Paul J. McMurdie.
Adapted for mia by Felix G.M. Ernst</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>esophagus</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioconductor.org/packages/scater/'>scater</a></span><span class='op'>)</span>
<span class='fu'>calculateUniFrac</span><span class='op'>(</span><span class='va'>esophagus</span>, weighted <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt;           B         C
#&gt; C 0.5175550          
#&gt; D 0.5182284 0.5422394</div><div class='input'><span class='fu'>calculateUniFrac</span><span class='op'>(</span><span class='va'>esophagus</span>, weighted <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt;           B         C
#&gt; C 0.2035424          
#&gt; D 0.2603371 0.2477016</div><div class='input'><span class='fu'>calculateUniFrac</span><span class='op'>(</span><span class='va'>esophagus</span>, weighted <span class='op'>=</span> <span class='cn'>TRUE</span>, normalized <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt;           B         C
#&gt; C 0.1050480          
#&gt; D 0.1401124 0.1422409</div><div class='input'><span class='co'># for using calculateUniFrac in conjunction with runMDS the tree argument</span>
<span class='co'># has to be given separately. In addition, subsetting using ntop must</span>
<span class='co'># be disabled</span>
<span class='va'>esophagus</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/runMDS.html'>runMDS</a></span><span class='op'>(</span><span class='va'>esophagus</span>, FUN <span class='op'>=</span> <span class='va'>calculateUniFrac</span>, name <span class='op'>=</span> <span class='st'>"UniFrac"</span>,
                    tree <span class='op'>=</span> <span class='fu'>rowTree</span><span class='op'>(</span><span class='va'>esophagus</span><span class='op'>)</span>,
                    exprs_values <span class='op'>=</span> <span class='st'>"counts"</span>,
                    ntop <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>esophagus</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'>reducedDim</span><span class='op'>(</span><span class='va'>esophagus</span><span class='op'>)</span>
</div><div class='output co'>#&gt;         [,1]         [,2]
#&gt; B  0.1699641  0.260500471
#&gt; C -0.3291116  0.005770947
#&gt; D  0.1591475 -0.266271417
#&gt; attr(,"eig")
#&gt; [1] 0.1625302 0.1387943 0.0000000
#&gt; attr(,"GOF")
#&gt; [1] 1 1</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Sudarshan A. Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


