<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculation prevalence information for features across samples — getPrevalence • mia</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculation prevalence information for features across samples — getPrevalence" />
<meta property="og:description" content="These functions calculate the population prevalence for taxonomic ranks in a
SummarizedExperiment-class object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/microbiome/mia/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculation prevalence information for features across samples</h1>
    <small class="dont-index">Source: <a href='https://github.com/microbiome/mia/blob/master/R/getPrevalence.R'><code>R/getPrevalence.R</code></a></small>
    <div class="hidden name"><code>getPrevalence.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These functions calculate the population prevalence for taxonomic ranks in a
<code>SummarizedExperiment-class</code> object.</p>
    </div>

    <pre class="usage"><span class='fu'>getPrevalence</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>getPrevalence</span><span class='op'>(</span><span class='va'>x</span>, detection <span class='op'>=</span> <span class='fl'>0</span>, include_lowest <span class='op'>=</span> <span class='cn'>FALSE</span>, sort <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>getPrevalence</span><span class='op'>(</span><span class='va'>x</span>, abund_values <span class='op'>=</span> <span class='st'>"counts"</span>, as_relative <span class='op'>=</span> <span class='cn'>TRUE</span>, rank <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>getPrevalentTaxa</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>getPrevalentTaxa</span><span class='op'>(</span><span class='va'>x</span>, prevalence <span class='op'>=</span> <span class='fl'>50</span><span class='op'>/</span><span class='fl'>100</span>, include_lowest <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>getPrevalentTaxa</span><span class='op'>(</span>
  <span class='va'>x</span>,
  prevalence <span class='op'>=</span> <span class='fl'>50</span><span class='op'>/</span><span class='fl'>100</span>,
  rank <span class='op'>=</span> <span class='cn'>NULL</span>,
  include_lowest <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='fu'>getRareTaxa</span><span class='op'>(</span><span class='va'>x</span>, rank <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>getRareTaxa</span><span class='op'>(</span><span class='va'>x</span>, rank <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>getPrevalentAbundance</span><span class='op'>(</span><span class='va'>x</span>, abund_values <span class='op'>=</span> <span class='st'>"relabundance"</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>getPrevalentAbundance</span><span class='op'>(</span><span class='va'>x</span>, abund_values <span class='op'>=</span> <span class='st'>"relabundance"</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>getPrevalentAbundance</span><span class='op'>(</span><span class='va'>x</span>, abund_values <span class='op'>=</span> <span class='st'>"counts"</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>agglomerateByPrevalence</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>agglomerateByPrevalence</span><span class='op'>(</span>
  <span class='va'>x</span>,
  rank <span class='op'>=</span> <span class='fu'><a href='taxonomy-methods.html'>taxonomyRanks</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1L</span><span class='op'>]</span>,
  other_label <span class='op'>=</span> <span class='st'>"Other"</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>SummarizedExperiment</a></code>
object</p></td>
    </tr>
    <tr>
      <th>detection</th>
      <td><p>Detection threshold for absence/presence. Either an
absolutes value compared directly to the values of <code>x</code> or a relative
value between 0 and 1, if <code>as_relative = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>include_lowest</th>
      <td><p>logical scalar: Should the lower boundary of the
detection and prevalence cutoffs be included? (default: <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>sort</th>
      <td><p>logical scalar: Should the result be sorted by prevalence?
(default: <code>FALSE</code>)</p></td>
    </tr>
    <tr>
      <th>abund_values</th>
      <td><p>A single character value for selecting the
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assay</a></code>
to use for prevalence calculation.</p></td>
    </tr>
    <tr>
      <th>as_relative</th>
      <td><p>logical scalar: Should the detection threshold be applied
on compositional (relative) abundances? (default: <code>TRUE</code>)</p></td>
    </tr>
    <tr>
      <th>rank, ...</th>
      <td><p>additional arguments</p><ul>
<li><p>If <code>!is.null(rank)</code> arguments are passed on to
<code><a href='agglomerate-methods.html'>agglomerateByRank</a></code>. See
<code><a href='agglomerate-methods.html'>?agglomerateByRank</a></code> for more details.</p></li>
<li><p>for <code>getPrevalentAbundance</code> additional parameters passed to
<code>getPrevalentTaxa</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>prevalence</th>
      <td><p>Prevalence threshold (in 0 to 1). The
required prevalence is strictly greater by default. To include the
limit, set <code>include_lowest</code> to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>other_label</th>
      <td><p>A single <code>character</code> valued used as the label for the
summary of non-prevalent taxa. (default: <code>other_label = "Other"</code>)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a named <code>numeric</code> vector. For <code>getPrevalence</code> the names are either
the row names of <code>x</code> or the names after agglomeration.</p>
<p>For <code>getPrevalentAbundance</code> the names correspond to the column name
names of <code>x</code> and include the joint abundance of prevalent taxa.</p>
<p>For <code>getPrevalentTaxa</code> only the names exceeding the threshold set by
<code>prevalence</code> are returned.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>getPrevalence</code> calculates the relative frequency of samples that exceed
the detection threshold. For <code>SummarizedExperiment</code> objects, the
prevalence is calculated for the selected taxonomic rank, otherwise for the
rows. The absolute population prevalence can be obtained by multiplying the
prevalence by the number of samples (<code><a href='https://rdrr.io/r/base/nrow.html'>ncol(x)</a></code>). If <code>as_relative =
TRUE</code> the relative frequency (between 0 and 1) is used to check against the
<code>detection</code> threshold.</p>
<p>The core abundance index from <code>getPrevalentAbundance</code> gives the relative
proportion of the core species (in between 0 and 1). The core taxa are
defined as those that exceed the given population prevalence threshold at the
given detection level as set for <code>getPrevalentTaxa</code>.</p>
<p><code>getPrevalentTaxa</code> returns taxa that are more prevalent with the
given detection threshold for the selected taxonomic rank.</p>
<p><code>getRareTaxa</code> returns complement of <code>getPrevalentTaxa</code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>A Salonen et al. The adult intestinal core microbiota is determined by
analysis depth and health status. Clinical Microbiology and Infection
18(S4):16 20, 2012.
To cite the R package, see citation('mia')</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='agglomerate-methods.html'>agglomerateByRank</a></code>,
<code><a href='getTopTaxa.html'>getTopTaxa</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Leo Lahti
For <code>getPrevalentAbundance</code>: Leo Lahti and Tuomas Borman.
Contact: <a href='microbiome.github.io'>microbiome.github.io</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>GlobalPatterns</span><span class='op'>)</span>
<span class='co'># Get prevalence estimates for individual ASV/OTU</span>
<span class='va'>prevalence.frequency</span> <span class='op'>&lt;-</span> <span class='fu'>getPrevalence</span><span class='op'>(</span><span class='va'>GlobalPatterns</span>,
                                      detection <span class='op'>=</span> <span class='fl'>0</span>,
                                      sort <span class='op'>=</span> <span class='cn'>TRUE</span>,
                                      as_relative <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>prevalence.frequency</span><span class='op'>)</span>
</div><div class='output co'>#&gt; 145149 114821 108747 526804 332405  98605 
#&gt;      1      1      1      1      1      1 </div><div class='input'>
<span class='co'># Get prevalence estimates for phylums</span>
<span class='co'># - the getPrevalence function itself always returns population frequencies</span>
<span class='va'>prevalence.frequency</span> <span class='op'>&lt;-</span> <span class='fu'>getPrevalence</span><span class='op'>(</span><span class='va'>GlobalPatterns</span>,
                                      rank <span class='op'>=</span> <span class='st'>"Phylum"</span>,
                                      detection <span class='op'>=</span> <span class='fl'>0</span>,
                                      sort <span class='op'>=</span> <span class='cn'>TRUE</span>,
                                      as_relative <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>prevalence.frequency</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     Chloroflexi           WPS-2      Firmicutes  Planctomycetes Verrucomicrobia 
#&gt;               1               1               1               1               1 
#&gt;             WS3 
#&gt;               1 </div><div class='input'>
<span class='co'># - to obtain population counts, multiply frequencies with the sample size,</span>
<span class='co'># which answers the question "In how many samples is this phylum detectable"</span>
<span class='va'>prevalence.count</span> <span class='op'>&lt;-</span> <span class='va'>prevalence.frequency</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>GlobalPatterns</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>prevalence.count</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     Chloroflexi           WPS-2      Firmicutes  Planctomycetes Verrucomicrobia 
#&gt;              26              26              26              26              26 
#&gt;             WS3 
#&gt;              26 </div><div class='input'>
<span class='co'># Detection threshold 1 (strictly greater by default);</span>
<span class='co'># Note that the data (GlobalPatterns) is here in absolute counts</span>
<span class='co'># (and not compositional, relative abundances)</span>
<span class='co'># Prevalence threshold 50 percent (strictly greater by default)</span>
<span class='va'>taxa</span> <span class='op'>&lt;-</span> <span class='fu'>getPrevalentTaxa</span><span class='op'>(</span><span class='va'>GlobalPatterns</span>,
                         rank <span class='op'>=</span> <span class='st'>"Phylum"</span>,
                         detection <span class='op'>=</span> <span class='fl'>1</span><span class='op'>/</span><span class='fl'>100</span>,
                         prevalence <span class='op'>=</span> <span class='fl'>50</span><span class='op'>/</span><span class='fl'>100</span>,
                         as_relative <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>taxa</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "Actinobacteria" "Proteobacteria" "Cyanobacteria"  "Bacteroidetes" 
#&gt; [5] "Firmicutes"    </div><div class='input'>
<span class='co'># getRareTaxa returns the inverse</span>
<span class='va'>rare</span> <span class='op'>&lt;-</span> <span class='fu'>getRareTaxa</span><span class='op'>(</span><span class='va'>GlobalPatterns</span>,
                    rank <span class='op'>=</span> <span class='st'>"Phylum"</span>,
                    detection <span class='op'>=</span> <span class='fl'>1</span><span class='op'>/</span><span class='fl'>100</span>,
                    prevalence <span class='op'>=</span> <span class='fl'>50</span><span class='op'>/</span><span class='fl'>100</span>,
                    as_relative <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>rare</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "Crenarchaeota" "Euryarchaeota" "Spirochaetes"  "MVP-15"       
#&gt; [5] "SBR1093"       "Fusobacteria" </div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>esophagus</span><span class='op'>)</span>
<span class='fu'>getPrevalentAbundance</span><span class='op'>(</span><span class='va'>esophagus</span>, abund_values <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;         B         C         D 
#&gt; 0.9605911 0.8980392 0.9086758 </div><div class='input'>
<span class='co'># data can be aggregated based on prevalent taxonomic results</span>
<span class='fu'>agglomerateByPrevalence</span><span class='op'>(</span><span class='va'>GlobalPatterns</span>,
                         rank <span class='op'>=</span> <span class='st'>"Phylum"</span>,
                         detection <span class='op'>=</span> <span class='fl'>1</span><span class='op'>/</span><span class='fl'>100</span>,
                         prevalence <span class='op'>=</span> <span class='fl'>50</span><span class='op'>/</span><span class='fl'>100</span>,
                         as_relative <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; class: TreeSummarizedExperiment 
#&gt; dim: 6 26 
#&gt; metadata(0):
#&gt; assays(1): counts
#&gt; rownames(6): Actinobacteria Proteobacteria ... Firmicutes Other
#&gt; rowData names(7): Kingdom Phylum ... Genus Species
#&gt; colnames(26): CL3 CC1 ... Even2 Even3
#&gt; colData names(7): X.SampleID Primer ... SampleType Description
#&gt; reducedDimNames(0):
#&gt; mainExpName: NULL
#&gt; altExpNames(0):
#&gt; rowLinks: NULL
#&gt; rowTree: NULL
#&gt; colLinks: NULL
#&gt; colTree: NULL</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Sudarshan Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


