<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import QIIME2 results to TreeSummarizedExperiment — loadFromQIIME2 • mia</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Import QIIME2 results to TreeSummarizedExperiment — loadFromQIIME2" />
<meta property="og:description" content="Results exported from QIMME2 can be imported as a
TreeSummarizedExperiment using loadFromQIIME2. Except for the
featureTableFile, the other data types, taxonomyTableFile,
refSeqFile and phyTreeFile, are optional, but are highly
encouraged to be provided." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/mia/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import QIIME2 results to <code>TreeSummarizedExperiment</code></h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/mia/blob/master/R/loadFromQIIME2.R'><code>R/loadFromQIIME2.R</code></a></small>
    <div class="hidden name"><code>loadFromQIIME2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Results exported from QIMME2 can be imported as a
<code>TreeSummarizedExperiment</code> using <code>loadFromQIIME2</code>. Except for the
<code>featureTableFile</code>, the other data types, <code>taxonomyTableFile</code>,
<code>refSeqFile</code> and <code>phyTreeFile</code>, are optional, but are highly
encouraged to be provided.</p>
    </div>

    <pre class="usage"><span class='fu'>loadFromQIIME2</span><span class='op'>(</span>
  <span class='va'>featureTableFile</span>,
  taxonomyTableFile <span class='op'>=</span> <span class='cn'>NULL</span>,
  sampleMetaFile <span class='op'>=</span> <span class='cn'>NULL</span>,
  featureNamesAsRefSeq <span class='op'>=</span> <span class='cn'>TRUE</span>,
  refSeqFile <span class='op'>=</span> <span class='cn'>NULL</span>,
  phyTreeFile <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>featureTableFile</th>
      <td><p>a single <code>character</code> value defining the file
path of the feature table to be imported.</p></td>
    </tr>
    <tr>
      <th>taxonomyTableFile</th>
      <td><p>a single <code>character</code> value defining the file
path of the taxonomy table to be imported. (default:
<code>taxonomyTableFile = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>sampleMetaFile</th>
      <td><p>a single <code>character</code> value defining the file path
of the sample metadata to be imported. The file has to be in tsv format.
(default: <code>sampleMetaFile = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>featureNamesAsRefSeq</th>
      <td><p><code>TRUE</code> or <code>FALSE</code>: Should the feature
names of the feature table be regarded as reference sequences? This setting
will be disregarded, if <code>refSeqFile</code> is not <code>NULL</code>. If the
feature names do not contain valid DNA characters only, the reference
sequences will not be set.</p></td>
    </tr>
    <tr>
      <th>refSeqFile</th>
      <td><p>a single <code>character</code> value defining the file path of
the reference sequences for each feature. (default: <code>refSeqFile =
  NULL</code>).</p></td>
    </tr>
    <tr>
      <th>phyTreeFile</th>
      <td><p>a single <code>character</code> value defining the file path of
the phylogenetic tree. (default: <code>phyTreeFile = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments:</p><ul>
<li><p><code>temp</code>: the temporary directory used for decompressing the
data. (default: <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir()</a></code>)</p></li>
<li><p><code>removeTaxaPrefixes</code>: <code>TRUE</code> or <code>FALSE</code>: Should
taxonomic prefixes be removed? (default:
<code>removeTaxaPrefixes = FALSE</code>)</p></li>
</ul></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A
<code><a href='https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-class.html'>TreeSummarizedExperiment</a></code>
object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Both arguments <code>featureNamesAsRefSeq</code> and <code>refSeqFile</code> can be used
to define reference sequences of features. <code>featureNamesAsRefSeq</code> is
only taken into account, if <code>refSeqFile</code> is <code>NULL</code>. No reference
sequences are tried to be created, if <code>featureNameAsRefSeq</code> is
<code>FALSE</code> and <code>refSeqFile</code> is <code>NULL</code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Bolyen E et al. 2019: Reproducible, interactive, scalable and extensible
microbiome data science using QIIME 2. Nature Biotechnology 37: 852–857.
<a href='https://doi.org/10.1038/s41587-019-0209-9'>https://doi.org/10.1038/s41587-019-0209-9</a></p>
<p><a href='https://qiime2.org'>https://qiime2.org</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='makeTreeSummarizedExperimentFromphyloseq.html'>makeTreeSummarizedExperimentFromphyloseq</a></code>
<code><a href='makeTreeSummarizedExperimentFromBiom.html'>makeTreeSummarizedExperimentFromBiom</a></code>
<code><a href='makeTreeSummarizedExperimentFromDADA2.html'>makeTreeSummarizedExperimentFromDADA2</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Yang Cao</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>featureTableFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"table.qza"</span>, package <span class='op'>=</span> <span class='st'>"mia"</span><span class='op'>)</span>
<span class='va'>taxonomyTableFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"taxonomy.qza"</span>, package <span class='op'>=</span> <span class='st'>"mia"</span><span class='op'>)</span>
<span class='va'>sampleMetaFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"sample-metadata.tsv"</span>, package <span class='op'>=</span> <span class='st'>"mia"</span><span class='op'>)</span>
<span class='va'>phyTreeFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"tree.qza"</span>, package <span class='op'>=</span> <span class='st'>"mia"</span><span class='op'>)</span>
<span class='va'>refSeqFile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"refseq.qza"</span>, package <span class='op'>=</span> <span class='st'>"mia"</span><span class='op'>)</span>
<span class='va'>tse</span> <span class='op'>&lt;-</span> <span class='fu'>loadFromQIIME2</span><span class='op'>(</span>
  featureTableFile <span class='op'>=</span> <span class='va'>featureTableFile</span>,
  taxonomyTableFile <span class='op'>=</span> <span class='va'>taxonomyTableFile</span>,
  sampleMetaFile <span class='op'>=</span> <span class='va'>sampleMetaFile</span>,
  refSeqFile <span class='op'>=</span> <span class='va'>refSeqFile</span>,
  phyTreeFile <span class='op'>=</span> <span class='va'>phyTreeFile</span>
<span class='op'>)</span>

<span class='va'>tse</span>
</div><div class='output co'>#&gt; class: TreeSummarizedExperiment 
#&gt; dim: 770 34 
#&gt; metadata(0):
#&gt; assays(1): counts
#&gt; rownames(770): 4b5eeb300368260019c1fbc7a3c718fc
#&gt;   fe30ff0f71a38a39cf1717ec2be3a2fc ... 98d250a339a635f20e26397dafc6ced3
#&gt;   1830c14ead81ad012f1db0e12f8ab6a4
#&gt; rowData names(8): Kingdom Phylum ... Species Confidence
#&gt; colnames(34): L1S8 L1S57 ... L6S68 L6S93
#&gt; colData names(9): sample.id barcode.sequence ...
#&gt;   reported.antibiotic.usage days.since.experiment.start
#&gt; reducedDimNames(0):
#&gt; mainExpName: NULL
#&gt; altExpNames(0):
#&gt; rowLinks: a LinkDataFrame (770 rows)
#&gt; rowTree: 1 phylo tree(s) (770 leaves)
#&gt; colLinks: NULL
#&gt; colTree: NULL
#&gt; referenceSeq: a DNAStringSet (770 sequences)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Sudarshan Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


