<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a phyloseq object from a TreeSummarizedExperiment object — makePhyloseqFromTreeSummarizedExperiment • mia</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create a phyloseq object from a TreeSummarizedExperiment object — makePhyloseqFromTreeSummarizedExperiment" />
<meta property="og:description" content="This function creates a phyloseq object from a TreeSummarizedExperiment
object. By using abund_values, it is possible to specify which table
from assay is added to the phyloseq object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/mia/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a phyloseq object from a TreeSummarizedExperiment object</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/mia/blob/master/R/makephyloseqFromTreeSummarizedExperiment.R'><code>R/makephyloseqFromTreeSummarizedExperiment.R</code></a></small>
    <div class="hidden name"><code>makePhyloseqFromTreeSummarizedExperiment.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function creates a phyloseq object from a TreeSummarizedExperiment
object. By using <code>abund_values</code>, it is possible to specify which table
from <code>assay</code> is added to the phyloseq object.</p>
    </div>

    <pre class="usage"><span class='fu'>makePhyloseqFromTreeSummarizedExperiment</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>makePhyloseqFromTreeSummarizedExperiment</span><span class='op'>(</span><span class='va'>x</span>, abund_values <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span>

<span class='co'># S4 method for TreeSummarizedExperiment</span>
<span class='fu'>makePhyloseqFromTreeSummarizedExperiment</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a <code>TreeSummarizedExperiment</code> object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments</p></td>
    </tr>
    <tr>
      <th>abund_values</th>
      <td><p>A single character value for selecting the
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assay</a></code> to be
included in the phyloseq object that is created. By default, it is counts
table.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code>Phyloseq</code> object.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>makePhyloseqFromTreeSummarizedExperiment</code> is used for creating a
phyloseq object from TreeSummarizedExperiment object.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Leo Lahti and Tuomas Borman. Contact: <a href='microbiome.github.io'>microbiome.github.io</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Get tse object</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"GlobalPatterns"</span><span class='op'>)</span>
<span class='va'>tse</span> <span class='op'>&lt;-</span> <span class='va'>GlobalPatterns</span>

<span class='co'># Create a phyloseq object from it</span>
<span class='va'>phy</span> <span class='op'>&lt;-</span> <span class='fu'>makePhyloseqFromTreeSummarizedExperiment</span><span class='op'>(</span><span class='va'>tse</span><span class='op'>)</span>
<span class='va'>phy</span>
</div><div class='output co'>#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
#&gt; sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
#&gt; phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</div><div class='input'>
<span class='co'># By default the chosen table is counts, but if there is other tables,</span>
<span class='co'># they can be chosen with abund_values.</span>

<span class='co'># Counts relative abundances table</span>
<span class='va'>tse</span> <span class='op'>&lt;-</span> <span class='fu'><a href='transformCounts.html'>transformCounts</a></span><span class='op'>(</span><span class='va'>tse</span>, method <span class='op'>=</span> <span class='st'>"relabundance"</span><span class='op'>)</span>
<span class='va'>phy2</span> <span class='op'>&lt;-</span> <span class='fu'>makePhyloseqFromTreeSummarizedExperiment</span><span class='op'>(</span><span class='va'>tse</span>, abund_values <span class='op'>=</span> <span class='st'>"relabundance"</span><span class='op'>)</span>
<span class='va'>phy2</span>
</div><div class='output co'>#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
#&gt; sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
#&gt; phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst, Sudarshan Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


