<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>mia datasets — mia-datasets • mia</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="mia datasets — mia-datasets"><meta property="og:description" content="These datasets are conversions of the phyloseq datasets
GlobalPatterns, enterotype, and esophagus into the
TreeSummarizedExperiment
data container.
dmn_se contains an example SummarizedExperiment derived
from data in the DirichletMultinomial package. See
?calculateDMN for more details.
Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample (2011)
This work compared the microbial communities from 25 environmental samples and three known ``mock communities''
at a an averag depth of 3.1 million reads per sample.
Authors reproduced diversity patterns seen in many other
published studies, while invesitigating technical issues/bias by
applying the same techniques to simulated microbial communities of known
composition. Many thanks to J. Gregory Caporaso for providing the OTU-clustered data files
for inclusion in the phyloseq package, from which this data TreeSummarizedExperiment
version was then converted.
The enterotype data of the human gut microbiome (Arumugam et al. 2011) includes
taxonomic profiling for 280 fecal samples from 22 subjects based on shotgun DNA sequencing.
The authors claimed that the data naturally clumps into three community-level clusters, or
``enterotypes'', that are not immediately explained by sequencing technology or demographic
features of the subjects. A later addendum (2014) the authors stated that enterotypes
&quot;should not be seen as discrete clusters, but as a way of stratifying samples to reduce complexity.&quot;
The esophagus data set from Pei et al. (2004)
includes 3 samples from 3 human adults based on biopsies analysed with 16S rDNA PCR.
The 16S rRNA sequence processing has been provided in the mothur wiki
at the link below.
This data set from Turnbaugh et al. (2009) was used to introduce
Dirichlet Multinomial Mixtures (DMM) for microbiota stratification by
Holmes et al. (2012).
PeerJ data by Potbhare et al. (2022) includes skin microbial profiles of 58 volunteers with multiple factors.
16S r-RNA sequencing of V3-V4 regions was done to generate millions of read using illumina platform.
A standard bioinformatic and statistical analysis done to explore skin bacterial diversity and its association with age, diet, geographical locations.
The authors investigated significant association of skin microbiota with individual’s geographical location.
The HintikkaXO dataset contains high-throughput profiling data from 40 rat
samples, including 39 biomarkers, 38 metabolites (NMR), and 12706 OTUs from
318 species, measured from Cecum. This is diet comparison study with High/Low
fat diet and xylo-oligosaccaride supplementation. Column metadata is common
for all experiments (microbiota, metabolites, biomarkers) and includes the
following fields:
Tengeler data by Tengeler et al. (2022) includes gut microbiota profiles of 27 persons with ADHD.
A standard bioinformatic and statistical analysis done to  demonstrate that altered microbial
composition could be a driver of altered brain structure and function and concomitant changes in the animals’ behavior.
They investigated this by colonizing young, male, germ-free C57BL/6JOlaHsd mice with microbiota from individuals with and without ADHD."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/microbiome/mia/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>mia datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/microbiome/mia/blob/HEAD/R/mia.R" class="external-link"><code>R/mia.R</code></a></small>
    <div class="hidden name"><code>mia-datasets.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These datasets are conversions of the <span class="pkg">phyloseq</span> datasets
<code>GlobalPatterns</code>, <code>enterotype</code>, and <code>esophagus</code> into the
<a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-class.html" class="external-link">TreeSummarizedExperiment</a>
data container.</p>
<p><code>dmn_se</code> contains an example <code>SummarizedExperiment</code> derived
from data in the <span class="pkg">DirichletMultinomial</span> package. See
<code><a href="calculateDMN.html">?calculateDMN</a></code> for more details.</p>
<p>Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample (2011)</p>
<p>This work compared the microbial communities from 25 environmental samples and three known ``mock communities''
at a an averag depth of 3.1 million reads per sample.
Authors reproduced diversity patterns seen in many other
published studies, while invesitigating technical issues/bias by
applying the same techniques to simulated microbial communities of known
composition. Many thanks to J. Gregory Caporaso for providing the OTU-clustered data files
for inclusion in the <span class="pkg">phyloseq</span> package, from which this data <code>TreeSummarizedExperiment</code>
version was then converted.</p>
<p>The enterotype data of the human gut microbiome (Arumugam et al. 2011) includes
taxonomic profiling for 280 fecal samples from 22 subjects based on shotgun DNA sequencing.
The authors claimed that the data naturally clumps into three community-level clusters, or
``enterotypes'', that are not immediately explained by sequencing technology or demographic
features of the subjects. A later addendum (2014) the authors stated that enterotypes
"should not be seen as discrete clusters, but as a way of stratifying samples to reduce complexity."</p>
<p>The esophagus data set from Pei et al. (2004)
includes 3 samples from 3 human adults based on biopsies analysed with 16S rDNA PCR.
The 16S rRNA sequence processing has been provided in the mothur wiki
at the link below.</p>
<p>This data set from Turnbaugh et al. (2009) was used to introduce
Dirichlet Multinomial Mixtures (DMM) for microbiota stratification by
Holmes et al. (2012).</p>
<p>PeerJ data by Potbhare et al. (2022) includes skin microbial profiles of 58 volunteers with multiple factors.
16S r-RNA sequencing of V3-V4 regions was done to generate millions of read using illumina platform.
A standard bioinformatic and statistical analysis done to explore skin bacterial diversity and its association with age, diet, geographical locations.
The authors investigated significant association of skin microbiota with individual’s geographical location.</p>
<p>The HintikkaXO dataset contains high-throughput profiling data from 40 rat
samples, including 39 biomarkers, 38 metabolites (NMR), and 12706 OTUs from
318 species, measured from Cecum. This is diet comparison study with High/Low
fat diet and xylo-oligosaccaride supplementation. Column metadata is common
for all experiments (microbiota, metabolites, biomarkers) and includes the
following fields:</p>
<p>Tengeler data by Tengeler et al. (2022) includes gut microbiota profiles of 27 persons with ADHD.
A standard bioinformatic and statistical analysis done to  demonstrate that altered microbial
composition could be a driver of altered brain structure and function and concomitant changes in the animals’ behavior.
They investigated this by colonizing young, male, germ-free C57BL/6JOlaHsd mice with microbiota from individuals with and without ADHD.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">esophagus</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dmn_se</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">peerj13075</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HintikkaXOData</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Tengeler2020</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>An object of class <code>TreeSummarizedExperiment</code> with 19216 rows and 26 columns.</p>
<p>An object of class <code>TreeSummarizedExperiment</code> with 553 rows and 280 columns.</p>
<p>An object of class <code>TreeSummarizedExperiment</code> with 58 rows and 3 columns.</p>
<p>An object of class <code>SummarizedExperiment</code> with 130 rows and 278 columns.</p>
<p>An object of class <code>TreeSummarizedExperiment</code> with 674 rows and 58 columns.</p>
<p>An object of class <code>MultiAssayExperiment</code> of length 3.</p>
<p>An object of class <code>TreeSummarizedExperiment</code> with 151 rows and 27 columns.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    
<ul><li><p>Sample: Sample ID (character)</p></li>
<li><p>Rat: Rat ID (factor)</p></li>
<li><p>Site: Site of measurement ("Cecum"); single value</p></li>
<li><p>Diet: Diet group (factor; combination of the Fat and XOS fields)</p></li>
<li><p>Fat: Fat in Diet (factor; Low/High)</p></li>
<li><p>XOS: XOS Diet Supplement (numeric; 0/1)</p></li>
</ul><p>Row metadata of the microbiota data contains taxonomic information on the
Phylum, Class, Order, Family, Genus, Species, and OTU levels.</p>
<p>Biomarker data contains 39 biomarkers.</p>
<p>Metabolite data contains 38 NMR metabolites.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Caporaso, J. G., et al. (2011).
Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample.
PNAS, 108, 4516-4522.
<a href="http://www.pnas.org/content/108/suppl.1/4516.short" class="external-link">http://www.pnas.org/content/108/suppl.1/4516.short</a></p>
<p>Arumugam, M., et al. (2011). Enterotypes of the human gut microbiome.
Nature, 473(7346), 174-180.
<a href="http://www.nature.com/doifinder/10.1038/nature09944" class="external-link">http://www.nature.com/doifinder/10.1038/nature09944</a>
Supplemental information includes subject data.
OTU-clustered data was initially downloaded from the publicly-accessible:
<a href="http://www.bork.embl.de/Docu/Arumugam_et_al_2011/downloads.html" class="external-link">http://www.bork.embl.de/Docu/Arumugam_et_al_2011/downloads.html</a></p>
<p>Arumugam, M., et al. (2014). Addendum: Enterotypes of the human gut microbiome.
Nature 506, 516 (2014). <a href="https://doi.org/10.1038/nature13075" class="external-link">https://doi.org/10.1038/nature13075</a></p>
<p>Pei, Z., Bini, E. J., Yang, L., Zhou, M., Francois, F., &amp; Blaser, M. J. (2004).
Bacterial biota in the human distal esophagus.
Proceedings of the National Academy of Sciences of the United States of America, 101(12), 4250-4255.
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC384727" class="external-link">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC384727</a></p>
<p>McMurdie, J. &amp; Holmes, S. (2013) <em>phyloseq</em>: An R Package for reproducible interactive analysis
and graphics of microbiome census data. PLoS ONE. 8(4):e61217.
<a href="https://doi.org/10.1371/journal.pone.0061217" class="external-link">https://doi.org/10.1371/journal.pone.0061217</a></p>
<p>Mothur-processed files and the sequence data can be downloaded at:
<a href="http://www.mothur.org/wiki/Esophageal_community_analysis" class="external-link">http://www.mothur.org/wiki/Esophageal_community_analysis</a></p>
<p>Holmes I, Harris K, Quince C (2012).
Dirichlet Multinomial Mixtures: Generative Models for Microbial Metagenomics.
PLoS ONE 7(2): e30126. <a href="https://doi.org/10.1371/journal.pone.0030126" class="external-link">https://doi.org/10.1371/journal.pone.0030126</a></p>
<p>Turnbaugh PJ, Hamady M, Yatsunenko T, Cantarel BL, Duncan A, et al. (2009).
A core gut microbiome in obese and lean twins. Nature 457: 480–484.
<a href="https://doi.org/10.1038/nature07540" class="external-link">https://doi.org/10.1038/nature07540</a></p>
<p>Potbhare, R., RaviKumar, A., Munukka, E., Lahti, L., &amp; Ashma, R. (2022).
Skin microbiota diversity among genetically unrelated individuals of Indian origin.
PeerJ, 10, e13075.
<a href="https://peerj.com/articles/13075/" class="external-link">https://peerj.com/articles/13075/</a>
Supplemental information includes OTU table and taxonomy table and publicly-accessible from:
<a href="https://www.doi.org/10.7717/peerj.13075/supp-1" class="external-link">https://www.doi.org/10.7717/peerj.13075/supp-1</a>
<a href="https://www.doi.org/10.7717/peerj.13075/supp-2" class="external-link">https://www.doi.org/10.7717/peerj.13075/supp-2</a></p>
<p>Hintikka L et al. (2021): Xylo-oligosaccharides in prevention of hepatic
steatosis and adipose tissue inflammation: associating taxonomic and
metabolomic patterns in fecal microbiotas with biclustering. International
Journal of Environmental Research and Public Health 18(8):4049
<a href="https://doi.org/10.3390/ijerph18084049" class="external-link">https://doi.org/10.3390/ijerph18084049</a></p>
<p>Tengeler, A.C., Dam, S.A., Wiesmann, M. et al.
Gut microbiota from persons with attention-deficit/hyperactivity disorder affects the brain in mice.
Microbiome 8, 44 (2020).
<a href="https://doi.org/10.1186/s40168-020-00816-x" class="external-link">https://doi.org/10.1186/s40168-020-00816-x</a>
Supplemental information includes Home-cage activity, methods, results and imaging parameters and publicly-accessible from:
<a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM1_ESM.docx" class="external-link">https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM1_ESM.docx</a>
<a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM2_ESM.docx" class="external-link">https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM2_ESM.docx</a>
<a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM3_ESM.docx" class="external-link">https://static-content.springer.com/esm/art%3A10.1186%2Fs40168-020-00816-x/MediaObjects/40168_2020_816_MOESM3_ESM.docx</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Caporaso, J. G., et al.</p>
<p>Arumugam, M., Raes, J., et al.</p>
<p>Pei et al. <a href="mailto:zhiheng.pei@med.nyu.edu">zhiheng.pei@med.nyu.edu</a>.</p>
<p>Turnbaugh, PJ et al.</p>
<p>Potbhare, R., et al.</p>
<p>Leo Lahti et al.</p>
<p>A.C. Tengeler, et al.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst, Sudarshan A. Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

