<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>mia: Microbiome analysis tools • mia</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="mia: Microbiome analysis tools">
<meta property="og:description" content="mia">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mia.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/microbiome/mia/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="mia_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>mia: Microbiome analysis tools</h1>
            
            <h4 class="date">2021-09-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/microbiome/mia/blob/master/vignettes/mia.Rmd"><code>vignettes/mia.Rmd</code></a></small>
      <div class="hidden name"><code>mia.Rmd</code></div>

    </div>

    
    
<p><code>mia</code> implements tools for microbiome analysis based on the <code>SummarizedExperiment</code> <span class="citation">(Morgan et al. 2020)</span>, <code>SingleCellExperiment</code> <span class="citation">(Amezquita et al. 2020)</span> and <code>TreeSummarizedExperiment</code> <span class="citation">(Huang 2021)</span> infrastructure. Data wrangling and analysis are the main scope of this package.</p>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install <code>mia</code>, install <code>BiocManager</code> first, if it is not installed. Afterwards use the <code>install</code> function from <code>BiocManager</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"mia"</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-mia" class="section level1">
<h1 class="hasAnchor">
<a href="#load-mia" class="anchor"></a>Load <em>mia</em>
</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/microbiome/mia">"mia"</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="loading-a-treesummarizedexperiment-object" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-a-treesummarizedexperiment-object" class="anchor"></a>Loading a <code>TreeSummarizedExperiment</code> object</h1>
<p>A few example datasets are available via <code>mia</code>. For this vignette the <code>GlobalPatterns</code> dataset is loaded first.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span>
<span class="va">se</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span>
<span class="va">se</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
</div>
<div id="functions-for-working-with-microbiome-data" class="section level1">
<h1 class="hasAnchor">
<a href="#functions-for-working-with-microbiome-data" class="anchor"></a>Functions for working with microbiome data</h1>
<p>One of the main topics for analysing microbiome data is the application of taxonomic data to describe features measured. The interest lies in the connection between individual bacterial species and their relation to each other.</p>
<p><code>mia</code> does not rely on a specific object type to hold taxonomic data, but uses specific columns in the <code>rowData</code> of a <code>SummarizedExperiment</code> object. <code>taxonomyRanks</code> can be used to construct a <code>character</code> vector of available taxonomic levels. This can be used, for example, for subsetting.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print the available taxonomic ranks</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># subset to taxonomic data only</span>
<span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="../reference/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## DataFrame with 19216 rows and 7 columns
##            Kingdom        Phylum        Class        Order        Family
##        &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;
## 549322     Archaea Crenarchaeota Thermoprotei           NA            NA
## 522457     Archaea Crenarchaeota Thermoprotei           NA            NA
## 951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423     Archaea Crenarchaeota        Sd-NA           NA            NA
## 586076     Archaea Crenarchaeota        Sd-NA           NA            NA
## ...            ...           ...          ...          ...           ...
## 278222    Bacteria           SR1           NA           NA            NA
## 463590    Bacteria           SR1           NA           NA            NA
## 535321    Bacteria           SR1           NA           NA            NA
## 200359    Bacteria           SR1           NA           NA            NA
## 271582    Bacteria           SR1           NA           NA            NA
##              Genus                Species
##        &lt;character&gt;            &lt;character&gt;
## 549322          NA                     NA
## 522457          NA                     NA
## 951     Sulfolobus Sulfolobusacidocalda..
## 244423          NA                     NA
## 586076          NA                     NA
## ...            ...                    ...
## 278222          NA                     NA
## 463590          NA                     NA
## 535321          NA                     NA
## 200359          NA                     NA
## 271582          NA                     NA</code></pre>
<p>The columns are recognized case insensitive. Additional functions are available to check for validity of taxonomic information or generate labels based on the taxonomic information.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/taxonomy-methods.html">taxonomyRankEmpty</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Species"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  1413 17803</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Class:Thermoprotei"               "Class:Thermoprotei_1"            
## [3] "Species:Sulfolobusacidocaldarius" "Class:Sd-NA"                     
## [5] "Class:Sd-NA_1"                    "Class:Sd-NA_2"</code></pre>
<p>For more details see the man page <code><a href="../reference/taxonomy-methods.html">?taxonomyRanks</a></code>.</p>
<div id="merging-and-agglomeration-based-on-taxonomic-information-" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-and-agglomeration-based-on-taxonomic-information-" class="anchor"></a>Merging and agglomeration based on taxonomic information.</h2>
<p>Agglomeration of data based on these taxonomic descriptors can be performed using functions implemented in <code>mia</code>. In addition to the <code>aggValue</code> functions provide by <code>TreeSummarizedExperiment</code> <code>agglomerateByRank</code> is available. <code>agglomerateByRank</code> does not require tree data to be present.</p>
<p><code>agglomerateByRank</code> constructs a <code>factor</code> to guide merging from the available taxonomic information. For more information on merging have a look at the man page via <code><a href="../reference/merge-methods.html">?mergeRows</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># agglomerate at the Family taxonomic rank</span>
<span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">se</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>
<span class="co">## How many taxa before/after agglomeration?</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 19216</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 603</code></pre>
<p>Tree data can also be shrunk alongside agglomeration, but this is turned of by default.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># with agglomeration of the tree</span>
<span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">se</span>, rank <span class="op">=</span> <span class="st">"Family"</span>,
                        agglomerateTree <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span> <span class="co"># same number of rows, but</span></code></pre></div>
<pre><code>## [1] 603</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowTree</span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="co"># ... different</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 19216 tips and 19215 internal nodes.
## 
## Tip labels:
##   549322, 522457, 951, 244423, 586076, 246140, ...
## Node labels:
##   , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...
## 
## Rooted; includes branch lengths.</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowTree</span><span class="op">(</span><span class="va">x2</span><span class="op">)</span> <span class="co"># ... tree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 496 tips and 260 internal nodes.
## 
## Tip labels:
##   Family:Cenarchaeaceae, Family:SAGMA-X, Family:Nitrososphaeraceae, Family:Sulfolobaceae, Family:Halobacteriaceae, Family:Methanobacteriaceae, ...
## Node labels:
##   root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>For <code>agglomerateByRank</code> to work, taxonomic data must be present. Even though only one rank is available for the <code>enterotype</code> dataset, agglomeration can be performed effectively de-duplicating entries for the genus level.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">enterotype</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span>
<span class="fu"><a href="../reference/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Genus"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 552 280 
## metadata(0):
## assays(1): counts
## rownames(552): -1 Prosthecochloris ... Syntrophococcus Mogibacterium
## rowData names(1): Genus
## colnames(280): AM.AD.1 AM.AD.2 ... TS98_V2 TS99.2_V2
## colData names(9): Enterotype Sample_ID ... Age ClinicalStatus
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>To keep data tidy, the agglomerated data can be stored as an alternative experiment in the object of origin. With this synchronized sample subsetting becomes very easy.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>, <span class="st">"family"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">x2</span></code></pre></div>
<p>Keep in mind, that if you set <code>na.rm = TRUE</code>, rows with <code>NA</code> or similar value (defined via the <code>empty.fields</code> argument) will be removed. Depending on these settings different number of rows will be returned.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">se</span>, rank <span class="op">=</span> <span class="st">"Species"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"species"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">se</span>, rank <span class="op">=</span> <span class="st">"Species"</span>, na.rm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 944  26</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"species"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 2307   26</code></pre>
<p>For convenience the function <code>splitByRanks</code> is available, which agglomerates data on all <code>ranks</code> selected. By default all available ranks will be used. The output is compatible to be stored as alternative experiments.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">altExps</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/splitByRanks.html">splitByRanks</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="va">se</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(7): Kingdom Phylum ... Genus Species
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">altExpNames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</code></pre>
</div>
<div id="constructing-a-tree-from-taxonomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-a-tree-from-taxonomic-data" class="anchor"></a>Constructing a tree from taxonomic data</h2>
<p>Constructing a taxonomic tree from taxonomic data stored in <code>rowData</code> is quite straightforward and uses mostly functions implemented in <code>TreeSummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Species"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="../reference/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">]</span>
<span class="va">taxa_res</span> <span class="op">&lt;-</span> <span class="fu">resolveLoop</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span><span class="op">)</span>
<span class="va">taxa_tree</span> <span class="op">&lt;-</span> <span class="fu">toTree</span><span class="op">(</span>data <span class="op">=</span> <span class="va">taxa_res</span><span class="op">)</span>
<span class="va">taxa_tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Species"</span><span class="op">)</span><span class="op">)</span>
<span class="va">rowNodeLab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Species"</span><span class="op">)</span>, make_unique <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Species"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">changeTree</span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Species"</span><span class="op">)</span>,
                                   rowTree <span class="op">=</span> <span class="va">taxa_tree</span>,
                                   rowNodeLab <span class="op">=</span> <span class="va">rowNodeLab</span><span class="op">)</span></code></pre></div>
</div>
<div id="transformation-of-assay-data" class="section level2">
<h2 class="hasAnchor">
<a href="#transformation-of-assay-data" class="anchor"></a>Transformation of assay data</h2>
<p>Transformation of count data stored in <code>assays</code> is also a main task when work with microbiome data. <code>transformCounts</code> can be used for this and offers a few choices of available transformations. A modified object is returned and the transformed counts are stored in a new <code>assay</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assayNames</span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "counts"</code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">anterotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transformCounts.html">transformCounts</a></span><span class="op">(</span><span class="va">enterotype</span>, method <span class="op">=</span> <span class="st">"log10"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu">assayNames</span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "counts"</code></pre>
<p>For more details have a look at the man page <code><a href="../reference/transformCounts.html">?transformCounts</a></code>.</p>
</div>
<div id="community-indices-" class="section level2">
<h2 class="hasAnchor">
<a href="#community-indices-" class="anchor"></a>Community indices.</h2>
<p>In the field of microbiome ecology several indices to describe samples and community of samples are available. In this vignette we just want to give a very brief introduction.</p>
<p>Functions for calculating alpha and beta diversity indices are available. Using <code>estimateDiversity</code> multiple diversity indices are calculated by default and results are stored automatically in <code>colData</code>. Selected indices can be calculated individually by setting <code>index = "shannon"</code> for example.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateDiversity.html">estimateDiversity</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">8</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "coverage"            "fisher"              "gini_simpson"       
## [4] "inverse_simpson"     "log_modulo_skewness" "shannon"            
## [7] "faith"</code></pre>
<p>Beta diversity indices are used to describe inter-sample connections. Technically they are calculated as <code>dist</code> object and reduced dimensions can be extracted using <code>cmdscale</code>. This is wrapped up in the <code>runMDS</code> function of the <code>scater</code> package, but can be easily used to calculated beta diversity indices using the established functions from the <code>vegan</code> package or any other package using comparable inputs.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Genus"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">runMDS</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">se</span>,<span class="st">"Genus"</span><span class="op">)</span>, 
                             FUN <span class="op">=</span> <span class="fu">vegan</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span>,
                             method <span class="op">=</span> <span class="st">"bray"</span>,
                             name <span class="op">=</span> <span class="st">"BrayCurtis"</span>, 
                             ncomponents <span class="op">=</span> <span class="fl">5</span>, 
                             exprs_values <span class="op">=</span> <span class="st">"counts"</span>, 
                             keep_dist <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>JSD and UniFrac are implemented in <code>mia</code> as well. <code>calculateJSD</code> can be used as a drop-in replacement in the example above (omit the <code>method</code> argument as well) to calculate the JSD. For calculating the UniFrac distance via <code>calculateUniFrac</code> either a <code>TreeSummarizedExperiment</code> must be used or a tree supplied via the <code>tree</code> argument. For more details see <code><a href="../reference/calculateJSD.html">?calculateJSD</a></code>, <code><a href="../reference/calculateUniFrac.html">?calculateUniFrac</a></code> or <code><a href="../reference/runDPCoA.html">?calculateDPCoA</a></code>.</p>
<p><code>runMDS</code> performs the decomposition. Alternatively <code>runNMDS</code> can also be used.</p>
</div>
<div id="other-indices" class="section level2">
<h2 class="hasAnchor">
<a href="#other-indices" class="anchor"></a>Other indices</h2>
<p><code>estimateDominance</code> and <code>estimateEvenness</code> implement other sample-wise indices. The function behave equivalently to <code>estimateDiversity</code>. For more information see the corresponding man pages.</p>
</div>
</div>
<div id="utility-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#utility-functions" class="anchor"></a>Utility functions</h1>
<p>To make migration and adoption as easy as possible several utility functions are available.</p>
<div id="data-loading-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#data-loading-functions" class="anchor"></a>Data loading functions</h2>
<p>Functions to load data from <code>biom</code> files, <code>QIIME2</code> output, <code>DADA2</code> objects <span class="citation">(Callahan et al. 2016)</span> or <code>phyloseq</code> objects are available.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">esophagus</span>, package <span class="op">=</span> <span class="st">"phyloseq"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">esophagus</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 58 taxa and 3 samples ]
## phy_tree()    Phylogenetic Tree: [ 58 tips and 57 internal nodes ]</code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">esophagus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeTreeSummarizedExperimentFromPhyloseq.html">makeTreeSummarizedExperimentFromPhyloseq</a></span><span class="op">(</span><span class="va">esophagus</span><span class="op">)</span>
<span class="va">esophagus</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 58 3 
## metadata(0):
## assays(1): counts
## rownames(58): 59_8_22 59_5_13 ... 65_9_9 59_2_6
## rowData names(0):
## colnames(3): B C D
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (58 rows)
## rowTree: 1 phylo tree(s) (58 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>For more details have a look at the man page, for examples <code><a href="../reference/makeTreeSummarizedExperimentFromPhyloseq.html">?makeTreeSummarizedExperimentFromPhyloseq</a></code>.</p>
</div>
<div id="general-wrapper-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#general-wrapper-functions" class="anchor"></a>General wrapper functions</h2>
<p><code>getAbundanceFeature</code> and <code>getAbundanceSample</code> are wrappers on row-wise or column-wise assay data subsetting.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbundance.html">getAbundanceSample</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"CC1"</span>, abund_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">abund</span> <span class="op">==</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"CC1"</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbundance.html">getAbundanceFeature</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"522457"</span>, abund_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">abund</span> <span class="op">==</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="st">"522457"</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="selecting-most-interesting-features" class="section level2">
<h2 class="hasAnchor">
<a href="#selecting-most-interesting-features" class="anchor"></a>Selecting most interesting features</h2>
<p><code>getTopTaxa</code> returns a vector of the most <code>top</code> abundant feature IDs.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summaries.html">getTopTaxa</a></span><span class="op">(</span><span class="va">se</span>,
                       method <span class="op">=</span> <span class="st">"mean"</span>,
                       top <span class="op">=</span> <span class="fl">5</span>,
                       abund_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="va">top_taxa</span></code></pre></div>
<pre><code>## [1] "549656" "331820" "279599" "360229" "317182"</code></pre>
</div>
<div id="generating-tidy-data" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-tidy-data" class="anchor"></a>Generating tidy data</h2>
<p>To generate tidy data as used and required in most of the tidyverse, <code>meltAssay</code> can be used. A <code>data.frame</code> in the long format will be returned.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">molten_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meltAssay.html">meltAssay</a></span><span class="op">(</span><span class="va">se</span>,
                         abund_values <span class="op">=</span> <span class="st">"counts"</span>,
                         add_row_data <span class="op">=</span> <span class="cn">TRUE</span>,
                         add_col_data <span class="op">=</span> <span class="cn">TRUE</span>
                         <span class="op">)</span>
<span class="va">molten_data</span></code></pre></div>
<pre><code>## # A tibble: 499,616 × 24
##    FeatureID SampleID counts Kingdom Phylum   Class   Order Family Genus Species
##    &lt;fct&gt;     &lt;fct&gt;     &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;  
##  1 549322    CL3           0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  2 549322    CC1           0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  3 549322    SV1           0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  4 549322    M31Fcsw       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  5 549322    M11Fcsw       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  6 549322    M31Plmr       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  7 549322    M11Plmr       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  8 549322    F21Plmr       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  9 549322    M31Tong       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
## 10 549322    M11Tong       0 Archaea Crenarc… Thermo… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
## # … with 499,606 more rows, and 14 more variables: X.SampleID &lt;fct&gt;,
## #   Primer &lt;fct&gt;, Final_Barcode &lt;fct&gt;, Barcode_truncated_plus_T &lt;fct&gt;,
## #   Barcode_full_length &lt;fct&gt;, SampleType &lt;fct&gt;, Description &lt;fct&gt;,
## #   coverage &lt;int&gt;, fisher &lt;dbl&gt;, gini_simpson &lt;dbl&gt;, inverse_simpson &lt;dbl&gt;,
## #   log_modulo_skewness &lt;dbl&gt;, shannon &lt;dbl&gt;, faith &lt;dbl&gt;</code></pre>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] phyloseq_1.37.0                scater_1.21.3                 
##  [3] ggplot2_3.3.5                  scuttle_1.3.1                 
##  [5] mia_1.1.14                     TreeSummarizedExperiment_2.1.4
##  [7] Biostrings_2.61.2              XVector_0.33.0                
##  [9] SingleCellExperiment_1.15.2    SummarizedExperiment_1.23.4   
## [11] Biobase_2.53.0                 GenomicRanges_1.45.0          
## [13] GenomeInfoDb_1.29.5            IRanges_2.27.2                
## [15] S4Vectors_0.31.3               BiocGenerics_0.39.2           
## [17] MatrixGenerics_1.5.4           matrixStats_0.60.1            
## [19] BiocStyle_2.21.3              
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0            colorspace_2.0-2           
##   [3] ellipsis_0.3.2              rprojroot_2.0.2            
##   [5] BiocNeighbors_1.11.0        fs_1.5.0                   
##   [7] ggrepel_0.9.1               bit64_4.0.5                
##   [9] fansi_0.5.0                 decontam_1.13.0            
##  [11] codetools_0.2-18            splines_4.1.0              
##  [13] sparseMatrixStats_1.5.3     cachem_1.0.5               
##  [15] knitr_1.33                  ade4_1.7-17                
##  [17] jsonlite_1.7.2              cluster_2.1.2              
##  [19] BiocManager_1.30.16         compiler_4.1.0             
##  [21] assertthat_0.2.1            Matrix_1.3-4               
##  [23] fastmap_1.1.0               lazyeval_0.2.2             
##  [25] cli_3.0.1                   BiocSingular_1.9.1         
##  [27] htmltools_0.5.1.1           tools_4.1.0                
##  [29] igraph_1.2.6                rsvd_1.0.5                 
##  [31] gtable_0.3.0                glue_1.4.2                 
##  [33] GenomeInfoDbData_1.2.6      reshape2_1.4.4             
##  [35] dplyr_1.0.7                 Rcpp_1.0.7                 
##  [37] jquerylib_0.1.4             pkgdown_1.6.1              
##  [39] rhdf5filters_1.5.0          vctrs_0.3.8                
##  [41] multtest_2.49.0             ape_5.5                    
##  [43] nlme_3.1-152                DECIPHER_2.21.0            
##  [45] iterators_1.0.13            DelayedMatrixStats_1.15.4  
##  [47] xfun_0.25                   stringr_1.4.0              
##  [49] beachmat_2.9.1              lifecycle_1.0.0            
##  [51] irlba_2.3.3                 zlibbioc_1.39.0            
##  [53] MASS_7.3-54                 scales_1.1.1               
##  [55] ragg_1.1.3                  biomformat_1.21.0          
##  [57] parallel_4.1.0              rhdf5_2.37.0               
##  [59] yaml_2.2.1                  memoise_2.0.0              
##  [61] gridExtra_2.3               sass_0.4.0                 
##  [63] stringi_1.7.3               RSQLite_2.2.7              
##  [65] desc_1.3.0                  foreach_1.5.1              
##  [67] ScaledMatrix_1.1.0          tidytree_0.3.4             
##  [69] permute_0.9-5               BiocParallel_1.27.4        
##  [71] rlang_0.4.11                pkgconfig_2.0.3            
##  [73] systemfonts_1.0.2           bitops_1.0-7               
##  [75] evaluate_0.14               lattice_0.20-44            
##  [77] Rhdf5lib_1.15.2             purrr_0.3.4                
##  [79] treeio_1.17.2               bit_4.0.4                  
##  [81] tidyselect_1.1.1            plyr_1.8.6                 
##  [83] magrittr_2.0.1              bookdown_0.22              
##  [85] R6_2.5.0                    generics_0.1.0             
##  [87] DelayedArray_0.19.2         DBI_1.1.1                  
##  [89] pillar_1.6.2                withr_2.4.2                
##  [91] mgcv_1.8-36                 survival_3.2-11            
##  [93] RCurl_1.98-1.3              tibble_3.1.3               
##  [95] crayon_1.4.1                utf8_1.2.2                 
##  [97] rmarkdown_2.10              viridis_0.6.1              
##  [99] grid_4.1.0                  data.table_1.14.0          
## [101] blob_1.2.2                  vegan_2.6-0                
## [103] digest_0.6.27               tidyr_1.1.3                
## [105] textshaping_0.3.5           munsell_0.5.0              
## [107] DirichletMultinomial_1.35.0 beeswarm_0.4.0             
## [109] viridisLite_0.4.0           vipor_0.4.5                
## [111] bslib_0.2.5.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-SCE" class="csl-entry">
Amezquita, Robert, Aaron Lun, Etienne Becht, Vince Carey, Lindsay Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span> <em>Nature Methods</em> 17: 137–45. <a href="https://www.nature.com/articles/s41592-019-0654-x">https://www.nature.com/articles/s41592-019-0654-x</a>.
</div>
<div id="ref-dada2" class="csl-entry">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. <span>“Dada2: High-Resolution Sample Inference from Illumina Amplicon Data.”</span> <em>Nature Methods</em> 13: 581–83. <a href="https://doi.org/10.1038/nmeth.3869">https://doi.org/10.1038/nmeth.3869</a>.
</div>
<div id="ref-TSE" class="csl-entry">
Huang, Ruizhu. 2021. <em>TreeSummarizedExperiment: TreeSummarizedExperiment: A S4 Class for Data with Tree Structures</em>.
</div>
<div id="ref-SE" class="csl-entry">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment">https://bioconductor.org/packages/SummarizedExperiment</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst, Sudarshan A. Shetty, Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
